[openldap] Add new port

GordonSmith opened the pull request:
Signed-off-by: Gordon Smith <GordonJSmith@gmail.com>

Add new port for OpenLDAP

- #### What does your PR fix?
  Fixes #18072

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  !windows Yes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

GordonSmith committed:
[openlap] Add new port

Merge pull request #1 from dg0yt/openldap

Signed-off-by: Gordon Smith <GordonJSmith@gmail.com>

Cheney-W commented on ports/openldap/portfile.cmake:
```diff
@@ -0,0 +1,59 @@
+)
+message("OS_RELEASE (OS_RELEASE_ERROR): ${OS_RELEASE} (${OS_RELEASE_ERROR})")
+
+if(OS_RELEASE STREQUAL "\"CentOS Linux\"")
+    vcpkg_extract_source_archive_ex(
+        OUT_SOURCE_PATH SOURCE_PATH
+        ARCHIVE ${ARCHIVE}
+        PATCHES
+        openssl.patch
+        m4.patch
+    )
+else()
+    vcpkg_extract_source_archive_ex(
+        OUT_SOURCE_PATH SOURCE_PATH
+        ARCHIVE ${ARCHIVE}
+        PATCHES
+        openssl.patch
+    )
+endif()
```
```suggestion
set(PATCH "openssl.patch")
if(OS_RELEASE STREQUAL "\"CentOS Linux\"")
    list(APPEND PATCH "m4.patch") 
endif()
vcpkg_extract_source_archive_ex(
    OUT_SOURCE_PATH SOURCE_PATH
    ARCHIVE ${ARCHIVE}
    PATCHES ${PATCH}
)
```

dg0yt replied:
Please make it easy to reason about the patches.
 - I wish there was a way to write `m4.patch` so that it can always be applied. In the current form, it will be tested only on CentOS systems. In that way, patch breakage may be unnoticed in CI on port updates.
 - I would also prefer that (most) patches are passed literally, instead of redirecting via variables. 

In addition, I assume that `"CentOS Linux"` isn't the right indicator here: What we really need might be to look for the version of some tool from the autotools family.

Last not least `${ARCHIVE}` should be quoted here.
~~~cmake
set(EXTRA_PATCHES "")
if(OS_RELEASE STREQUAL [["CentOS Linux"]])
    list(APPEND EXTRA_PATCHES "m4.patch") 
endif()

vcpkg_extract_source_archive_ex(
    OUT_SOURCE_PATH SOURCE_PATH
    ARCHIVE "${ARCHIVE}"
    PATCHES
        openssl.patch
        ${EXTRA_PATCHES}
)
~~~

GordonSmith replied:
The m4.patch was only needed to force a pkg-config rerun (as the OpenLDAP pre-canned version was generated with a newer version of pkg-config than CentOS has) - there may well be a simpler way to do this, but I am new to autoconf, configure and pkg-config and just followed the error messages...
(IOW suggestions welcome!!!)

dg0yt replied:
I'm not sure what you mean by "pkg-config rerun". (You can trigger `autoreconf` via the `AUTOCONFIG` option `vcpkg_configure_make`. But it may cause other issues.)

GordonSmith replied:
I have removed the need for the m4.patch

Cheney-W commented on ports/openldap/portfile.cmake:
```diff
@@ -0,0 +1,59 @@
+    "LIBS=-ldl"
+)
+
+vcpkg_build_make(BUILD_TARGET depend)
+vcpkg_build_make()
+vcpkg_install_make()
```
```suggestion
vcpkg_install_make()
```
`vcpkg_install_make()` will automatically call `vcpkg_build_make()`, if `BUILD_TARGET` is not passed, the default is all.

dg0yt replied:
And `vcpkg_install_make` will overwrite the build logs from `vcpkg_build_make`.

GordonSmith replied:
vcpkg_build_make(BUILD_TARGET depend) will still needed to be called prior to vcpkg_install_make(), but I will remove the extra vcpkg_build_make()

GordonSmith replied:
I have removed the second call to vcpk_build_make and added the LOGFILE setting

dg0yt commented on ports/openldap/openssl.patch:
```diff
@@ -0,0 +1,13 @@
+ 
+ 		AC_CHECK_LIB(ssl, SSL_export_keying_material_early,
+-			[have_openssl=yes], [have_openssl=no],
++			[have_openssl=yes], [have_openssl=yes],
```
Do we need this patch?
- Can't we just pass a cache value via `OPTIONS`, instead of patching?
- If we are concerned about `ssl` not being found despite proper dependencies, shouldn't it be turned into a hard error?

GordonSmith replied:
OpenLDAP is locating the vcpkg version of openssl and checks if its version is new enough and all is good (I am passing in the appropriate flags to configure it withsll).

The problem is that it then goes on to check for the presence of SSL_export_keying_material_early in the SSL library, which fails as it is a static library (as that function is indeed present),

dg0yt replied:
> The problem is that it then goes on to check for the presence of SSL_export_keying_material_early in the SSL library, which fails as it is a static library (as that function is indeed present),

IIUC it fails for static linkage because the test doesn't add the additional libs needed to resolve symbols in libssl. This error would be resolved by passing them in `LIBS` or by patching the last parameter. 
But if we can rely on the function being present, we can just pass `ac_cv_lib_ssl_SSL_export_keying_material_early=yes` to with `OPTIONS` to provide the known result as if it was cached. (Cf. https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Libraries.html, and ...config.log for actual `ac_cv_...` variables.)

GordonSmith replied:
I fixed the patch so the test now succeeds (rather than ignored)

dg0yt commented on ports/openldap/portfile.cmake:
```diff
@@ -0,0 +1,59 @@
+    SHA512 30fdc884b513c53169910eec377c2ad05013b9f06bab3123d50d028108b24548791f7f47f18bcb3a2b4868edeab02c10d81ffa320c02d7b562f2e8f2fa25d6c9
+)
+
+execute_process(COMMAND bash "-c" "awk -F= '/^NAME/{print $2}' \"/etc/os-release\""
```
Does this need to go through bash, or can it be written as:
~~~cmake
 execute_process(COMMAND awk -F= [[/^NAME/{print $2}]] /etc/os-release
~~~

GordonSmith replied:
I have removed the need for an OS check

dg0yt commented on ports/openldap/portfile.cmake:
```diff
@@ -0,0 +1,59 @@
+    ERROR_VARIABLE OS_RELEASE_ERROR
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+)
+message("OS_RELEASE (OS_RELEASE_ERROR): ${OS_RELEASE} (${OS_RELEASE_ERROR})")
```
IMO this is not needed. If not removed, it needs a message type keyword, e.g. `STATUS`.

GordonSmith replied:
This debug message has been removed

dg0yt commented on ports/openldap/portfile.cmake:
```diff
@@ -0,0 +1,59 @@
+    SOURCE_PATH "${SOURCE_PATH}"
+    AUTOCONFIG
+
+    OPTIONS
+    ${BUILD_OPTS}
+    --disable-slapd
+    --with-tls=openssl
+    --without-cyrus-sasl
+    "LIBS=-ldl"
```
~~~
    OPTIONS
        ${BUILD_OPTS}
        --disable-slapd
        --with-tls=openssl
        --without-cyrus-sasl
        "LIBS=-ldl"

GordonSmith replied:
Fixed indentation and removed the need for the m4.patch

dg0yt commented on ports/openldap/portfile.cmake:
```diff
@@ -0,0 +1,59 @@
+    )
+endif()
+
+if(VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
+    set(BUILD_OPTS --enable-shared=yes --enable-static=no)
+else()
+    set(BUILD_OPTS --enable-shared=no --enable-static=yes)
+endif()
```
This is already handled by `vcpkg_configure_make`.

GordonSmith replied:
Removed unneeded code.

dg0yt commented on ports/openldap/portfile.cmake:
```diff
@@ -0,0 +1,59 @@
+    "LIBS=-ldl"
+)
+
+vcpkg_build_make(BUILD_TARGET depend)
```
Does this target need to be called explicitly in advance?

GordonSmith replied:
Yes

dg0yt commented on ports/openldap/vcpkg.json:
```diff
@@ -0,0 +1,15 @@
+  "description": "OpenLDAP Software is an open source implementation of the Lightweight Directory Access Protocol.",
+  "homepage": "https://www.openldap.org/software/",
+  "license": "OLDAP-2.8",
+  "supports": "!windows",
```
Doesn't it support mingw at least?

GordonSmith replied:
I added mingw to the supports - will see what the GH actions here does with it!
I made the logfile unique

dg0yt commented on ports/openldap/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+    OUT_SOURCE_PATH SOURCE_PATH
+    ARCHIVE "${ARCHIVE}"
+    PATCHES
+        openssl.patch
```
I verified that my proposed alternative (`ac_cv_...`) works.

GordonSmith replied:
I fixed the actual issue that was causing AC_CHECK_LIB to fail.

dg0yt commented on ports/openldap/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+        openssl.patch
+)
+
+execute_process(
+    COMMAND aclocal
+    COMMAND libtoolize
+    COMMAND autoreconf
+    WORKING_DIRECTORY ${SOURCE_PATH}
+)
```
1. Pristine `execute_process` is rarely used in vcpkg portfiles. (There is a special maintainer function which handles log files and `--only-downloads` mode.)
2. IMO this is not needed. I added `AUTOCONFIG` to `vcpkg_configure_make`, and the log shows that `autoreconf` does run the other tools:
~~~
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force 
autoreconf: configure.ac: tracing
autoreconf: running: libtoolize --copy --force
autoreconf: running: /usr/bin/autoconf --force
autoreconf: running: /usr/bin/autoheader --force
autoreconf: configure.ac: not using Automake
autoreconf: Leaving directory `.'
~~~

GordonSmith replied:
@dg0yt the subtle difference appears to be the call to "autoreconf" prior to calling "autoconf".  Here is the error message from your PR on CentOS-7:
```
ar: creating liblutil.a
libtool: Version mismatch error.  This is libtool 2.4.2, but the
libtool: definition of this LT_INIT comes from libtool 2.4.6.
libtool: You should recreate aclocal.m4 with macros from libtool 2.4.2
libtool: and run autoconf again.
make[2]: *** [Makefile:393: assert.lo] Error 63
make[2]: *** Waiting for unfinished jobs....
libtool: Version mismatch error.  This is libtool 2.4.2, but the
libtool: definition of this LT_INIT comes from libtool 2.4.6.
libtool: You should recreate aclocal.m4 with macros from libtool 2.4.2
libtool: and run autoconf again.
make[2]: *** [Makefile:393: decode.lo] Error 63
libtool: Version mismatch error.  This is libtool 2.4.2, but the
libtool: definition of this LT_INIT comes from libtool 2.4.6.
libtool: You should recreate aclocal.m4 with macros from libtool 2.4.2
libtool: and run autoconf again.
make[2]: *** [Makefile:393: encode.lo] Error 63
libtool: Version mismatch error.  This is libtool 2.4.2, but the
libtool: definition of this LT_INIT comes from libtool 2.4.6.
libtool: You should recreate aclocal.m4 with macros from libtool 2.4.2
libtool: and run autoconf again.
libtool: Version mismatch error.  This is libtool 2.4.2, but the
libtool: definition of this LT_INIT comes from libtool 2.4.6.
libtool: You should recreate aclocal.m4 with macros from libtool 2.4.2
libtool: and run autoconf again.
make[2]: *** [Makefile:393: io.lo] Error 63
make[2]: *** [Makefile:393: bprint.lo] Error 63
libtool: Version mismatch error.  This is libtool 2.4.2, but the
libtool: definition of this LT_INIT comes from libtool 2.4.6.
libtool: You should recreate aclocal.m4 with macros from libtool 2.4.2
libtool: and run autoconf again.
make[2]: *** [Makefile:393: options.lo] Error 63
libtool: Version mismatch error.  This is libtool 2.4.2, but the
libtool: definition of this LT_INIT comes from libtool 2.4.6.
libtool: You should recreate aclocal.m4 with macros from libtool 2.4.2
libtool: and run autoconf again.
libtool: Version mismatch error.  This is libtool 2.4.2, but the
libtool: definition of this LT_INIT comes from libtool 2.4.6.
libtool: You should recreate aclocal.m4 with macros from libtool 2.4.2
libtool: and run autoconf again.
make[2]: *** [Makefile:393: debug.lo] Error 63
libtool: Version mismatch error.  This is libtool 2.4.2, but the
libtool: definition of this LT_INIT comes from libtool 2.4.6.
libtool: You should recreate aclocal.m4 with macros from libtool 2.4.2
libtool: and run autoconf again.
make[2]: *** [Makefile:393: sockbuf.lo] Error 63
make[2]: *** [Makefile:393: memory.lo] Error 63
libtool: Version mismatch error.  This is libtool 2.4.2, but the
libtool: definition of this LT_INIT comes from libtool 2.4.6.
libtool: You should recreate aclocal.m4 with macros from libtool 2.4.2
libtool: and run autoconf again.
make[2]: *** [Makefile:393: stdio.lo] Error 63
make[1]: *** [Makefile:313: all-common] Error 1
make: *** [Makefile:321: all-common] Error 1
```

Having said that out loud, I just tested the following and it appears to works:
```
execute_process(
    COMMAND autoreconf
    WORKING_DIRECTORY ${SOURCE_PATH}
)

vcpkg_configure_make(
    SOURCE_PATH "${SOURCE_PATH}"
    OPTIONS
        --disable-slapd
        --with-tls=openssl
        --without-cyrus-sasl
        "LIBS=-ldl"
)
```

Again thanks for your help with this - its certainly a lot cleaner now than it was...

dg0yt replied:
> I just tested the following and it appears to works:

And this is what the `AUTOCONFIG` option is doing...

Can you please wait for my PR, so I don't get (more) merge conflicts? I switched to pkg-config for openssl, and I do have a few more changes.

dg0yt reviewed (changes requested):
I'm testing your PR locally, and I want to sent you a PR with proposed changed.

Cheney-W commented on ports/openldap/vcpkg.json:
```diff
@@ -0,0 +1,15 @@
+  "license": "OLDAP-2.8",
+  "supports": "!windows, (mingw & !x86)",
+  "dependencies": [
+    "openssl",
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    }
```
```suggestion
    "openssl"
```
The `vcpkg_cmake_xx` function is not used in portfile.cmake.

GordonSmith replied:
Removed

dg0yt replied:
You need to remove it again. I missed that.

dg0yt commented on ports/openldap/portfile.cmake:
```diff
@@ -0,0 +1,40 @@
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/var")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/var")
+
+file(INSTALL "${CMAKE_CURRENT_LIST_DIR}/FindOpenLDAP.cmake" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
```
Oh no. No unofficial find modules, unless entirely in unofficial namespaces.

GordonSmith replied:
Fixed in your PR

dg0yt commented on ports/openldap/usage:
```diff
@@ -0,0 +1,5 @@
+The package OpenLDAP provides CMake variables:
+
+    find_package(OpenLDAP REQUIRED)
+    target_include_directories(main PRIVATE ${OPENLDAP_INCLUDE_DIR})
+    target_link_libraries(main PRIVATE ${OPENLDAP_LIBRARIES})
```
The official usage for openldap is via pkg-config. This can be used in CMake, too.

GordonSmith replied:
@dg0yt - thanks, I will track down an example of this pattern in other ports.

GordonSmith replied:
Thanks for your PR for this.

dg0yt reviewed (changes requested):
From my PR in https://github.com/GordonSmith/vcpkg/pull/1, I ask for the following changes to be also included.
 - Use pkg-config for openssl.
 - Use `AUTOCONFIG` instead of running `autoreconf` separately.
 - No `LIBS=-ldl` without good reason.

GordonSmith commented:
@dg0yt I had missed your last comments on my repo...
* I have now included your pkg-config for openssl changes
* I still need to run autoreconf when libtool version != 2.4.6 - so I added a cmake check for that specific case - at least its more obvious why its needed?
* No longer need the `LIBS=-ldl` in CentOS - is this because your pkg-config tweak effectivly resolves them?

Thx.

GordonSmith commented:
Note:  I tweaked the configure parameters to remove all the warnings generated in buildtrees/openldap/config-x64-linux-dbg-err.log

dg0yt commented:
https://github.com/microsoft/vcpkg/blob/52b66042533aaf7f5721941941fd12fbf46164d4/scripts/cmake/vcpkg_configure_make.cmake#L540-L558

dg0yt commented:
> * I still need to run autoreconf when libtool version != 2.4.6 - so I added a cmake check for that specific case - at least its more obvious why its needed?

I didn't say it isn't needed. I said repeatedly that passing `AUTOCONFIG` to `vcpkg_configure_make` will do just that: Run `autoreconf` before running `configure`.

> * No longer need the `LIBS=-ldl` in CentOS - is this because your pkg-config tweak effectivly resolves them?

Probably yes: But that's what I already said in another comment. 
(I spent much time on fixing pc files.)

GordonSmith commented:
Unfortunately calling `autoreconf -i` on CentOS-7 will break the build, while `autoreconf -f` or `autoreconf` does not.
And using AUTOCONFIG is not needed when libtool version == 2.4.6 

Output log when using `autoreconf -vfi` option:
```
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force 
autoreconf: configure.ac: tracing
autoreconf: running: libtoolize --copy --force
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, `build'.
libtoolize: copying file `build/ltmain.sh'
libtoolize: You should add the contents of the following files to `aclocal.m4':
libtoolize:   `/usr/share/aclocal/libtool.m4'
libtoolize:   `/usr/share/aclocal/ltoptions.m4'
libtoolize:   `/usr/share/aclocal/ltversion.m4'
libtoolize:   `/usr/share/aclocal/ltsugar.m4'
libtoolize:   `/usr/share/aclocal/lt~obsolete.m4'
libtoolize: Consider adding `AC_CONFIG_MACRO_DIR([m4])' to configure.ac and
libtoolize: rerunning libtoolize, to keep the correct libtool macros in-tree.
libtoolize: Consider adding `-I m4' to ACLOCAL_AMFLAGS in Makefile.am.
autoreconf: running: /usr/bin/autoconf --force
autoreconf: running: /usr/bin/autoheader --force
autoreconf: configure.ac: not using Automake
autoreconf: Leaving directory `.'
```

Output log when using `autoreconf -v` option:
```
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal 
autoreconf: configure.ac: tracing
autoreconf: configure.ac: not running libtoolize: --install not given
autoreconf: running: /usr/bin/autoconf
autoreconf: running: /usr/bin/autoheader
autoreconf: configure.ac: not using Automake
autoreconf: Leaving directory `.'
```

dg0yt commented:
What kind of breakage happens with `autoreconf -vfi`?
What about the proposal to add `AC_CONFIG_MACRO_DIR([m4])`?

GordonSmith commented:
That is what my very first PR did

GordonSmith commented:
@dy0yt - I added the m4 patch into the openssl.patch and suspect we have a winner!

dg0yt reviewed (dismissed):
Great!
Maybe it is better to not squash too often/too early. The repo maintainers squash the PR with the merge anyways.

GordonSmith commented:
I normally do that - but I thought the bots here didn't like that?

dg0yt commented:
> I normally do that - but I thought the bots here didn't like that?

There is only one relevant bot, checking that the git tree object referenced in `versions` does actually exist. It won't complain as long as you update `versions` (every time) after committing port changes. It is easy to miss this update after amending a PR :-)

GordonSmith committed:
[openlap] Add new port

m4.patch is needed when libtool < 2.4.6
m4.patch will only work with autoconf < 2.70

Signed-off-by: Gordon Smith <GordonJSmith@gmail.com>

GordonSmith commented:
@dg0yt Another tweak to fix OSX build - FYI the m4.patch won't work work with autoconf version >= 2.70 - looks like this ties in with libtool = 2.4.6 which doesn't need the m4 patch.

dg0yt commented on ports/openldap/portfile.cmake:
```diff
@@ -0,0 +1,60 @@
+
+vcpkg_list(SET EXTRA_PATCHES)
+
+# Check autoconf version < 2.70
+execute_process(COMMAND autoconf --version OUTPUT_VARIABLE AUTOCONF_VERSION_STR)
+if(NOT "${AUTOCONF_VERSION_STR}" STREQUAL "" AND "${AUTOCONF_VERSION_STR}" MATCHES ".*2\\.[0-6].*")
+    vcpkg_list(APPEND EXTRA_PATCHES m4.patch)
+endif()
```
I'd prefer it the patch could be rewritten in a way so that it always applies. If it is only used on some systems, it may be ignored by CI and bitrot.

GordonSmith replied:
I agree - unfortunately I have zero experience with configure and pkg-config...

The reason the m4 patch is being added is to trick the autoconfig -vfi to work.  The other option is to remove the AUTOCONFIG and replace with a manual call to:  `autoreconf -vf` and then there is no need to for the m4 patch.

GordonSmith replied:
Making the portfile.cmake "clean" for libtool=2.4.6 breaks it for 2.4.2 and vice versa (the suggestion to add `AC_CONFIG_MACRO_DIR([m4])` doesn't work when autoconf >= 2.70)

GordonSmith committed:
wip:  Make cyrus-sasl optional

BillyONeal reviewed (approved):
I think this is fine but would like to ensure @dg0yt is happy before merging.

New port checklist:
* Review the code ✅
* Check the name against https://repology.org/ ✅
* Check the name against Bing/Google. (If the first results aren't C++ related, try "name C++") ✅
* Check the source code for optional `find_package`s ✅ (not a CMake port)
* Check that the versioning scheme and license match what upstream says✅
* Check that the source code comes from the upstream project's authoritative source.✅
* Check that the generated usage is accurate. ✅ (I'm assuming that it's OK given that explicit usage is provided and looks reasonable)
* Check that the reported license is accurate.✅

BillyONeal commented:
@dg0yt Can you confirm that you're happy? If I don't hear by tomorrow morning I will merge this.

BillyONeal commented:
Thanks!
